src/libstd/collections/hash/map.rs

```diff
@@ -1676,21 +1784,598 @@ impl<'a, K, V> InternalEntry<K, V, &'a mut RawTable<K, V>> {
+    pub fn hash_by<Q: ?Sized>(self, k: &Q) -> RawImmutableEntryBuilderHashed<'a, K, V>
+        where Q: Hash
+    {
+        self.hash_with(|mut hasher| {
+            k.hash(&mut hasher);
+            hasher.finish()
+        })
+    }
+
+    /// Initializes the raw entry builder with the hash yielded by the given function.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn hash_with<F>(self, func: F) -> RawImmutableEntryBuilderHashed<'a, K, V>
+        where F: FnOnce(S::Hasher) -> u64
+    {
+        let hasher = self.map.hash_builder.build_hasher();
+        let hash = SafeHash::new(func(hasher));
+
+        RawImmutableEntryBuilderHashed { map: &self.map.table, hash }
+    }
+
+    /// Searches for the location of the raw entry with the given query value.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn search_by<Q: ?Sized>(self, k: &Q) -> Option<(&'a K, &'a V)>
+        where K: Borrow<Q>,
+              Q: Eq + Hash
+    {
+        self.hash_by(k).search_by(k)
+    }
+}
+
+impl<'a, K, V> RawImmutableEntryBuilderHashed<'a, K, V>
+{
+    /// Searches for the location of the raw entry with the given query value.
+    ///
+    /// Note that it isn't required that the query value be hashable, as the
+    /// builder's hash is used.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn search_by<Q: ?Sized>(self, k: &Q) -> Option<(&'a K, &'a V)>
+        where K: Borrow<Q>,
+              Q: Eq,
+    {
+        self.search_with(|key| key.borrow() == k)
+    }
+
+    /// Searches for the location of the raw entry with the given comparison function.
+    ///
+    /// Note that mutable access is given to each key that is visited, because
+    /// this land is truly godless, and *someone* might have a use for this.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn search_with<F>(self, func: F) -> Option<(&'a K, &'a V)>
+        where F: FnMut(&K) -> bool,
+    {
+        match search_hashed(self.map, self.hash, func) {
+            InternalEntry::Occupied { elem } => {
+                Some(elem.into_refs())
+            }
+            InternalEntry::Vacant { .. } | InternalEntry::TableIsEmpty  => {
+                None
+            }
+        }
+    }
+}
+
+impl<'a, K, V> RawEntry<'a, K, V> {
+    /// Ensures a value is in the entry by inserting the default if empty, and returns
+    /// a mutable reference to the value in the entry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map: HashMap<&str, u32> = HashMap::new();
+    /// map.entry("poneyland").or_insert(12);
+    ///
+    /// assert_eq!(map["poneyland"], 12);
+    ///
+    /// *map.entry("poneyland").or_insert(12) += 10;
+    /// assert_eq!(map["poneyland"], 22);
+    /// ```
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn or_insert(self, default_key: K, default_val: V) -> (&'a mut K, &'a mut V) {
```

Comment:
FWIW `Entry::or_insert` only returns `&mut V` so this is somewhat inconsistent with that, albeit more general

Reply:
that's definitely an intentional design decision, but I could be convinced to go back